processing/src/main/java/io/druid/query/expression/TimestampFormatExprMacro.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.expression;
+
+import com.google.common.base.Preconditions;
+import io.druid.java.util.common.IAE;
+import io.druid.math.expr.Expr;
+import io.druid.math.expr.ExprEval;
+import io.druid.math.expr.ExprMacroTable;
+import org.joda.time.DateTimeZone;
+import org.joda.time.format.DateTimeFormat;
+import org.joda.time.format.DateTimeFormatter;
+import org.joda.time.format.ISODateTimeFormat;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+
+public class TimestampFormatExprMacro implements ExprMacroTable.ExprMacro
+{
+  @Override
+  public String name()
+  {
+    return "timestamp_format";
+  }
+
+  @Override
+  public Expr apply(final List<Expr> args)
+  {
+    if (args.size() < 1 || args.size() > 3) {
+      throw new IAE("Function[%s] must have 1 to 3 arguments", name());
+    }
+
+    final Expr arg = args.get(0);
+    final String formatString;
+    final DateTimeZone timeZone;
+
+    if (args.size() > 1) {
+      Preconditions.checkArgument(args.get(1).isLiteral(), "Function[%s] format arg must be a literal", name());
+      formatString = (String) args.get(1).getLiteralValue();
+    } else {
+      formatString = null;
+    }
+
+    if (args.size() > 2) {
+      timeZone = ExprUtils.toTimeZone(args.get(2));
+    } else {
+      timeZone = DateTimeZone.UTC;
+    }
+
+    final DateTimeFormatter formatter = formatString == null
+                                        ? ISODateTimeFormat.dateTime()
+                                        : DateTimeFormat.forPattern(formatString).withZone(timeZone);
+
+    class TimestampFormatExpr implements Expr
```

Comment:
> If something is a common pattern is Druid it doesn't mean that it's a good thing. Actually infinite methods like QueryableIndexStorageAdapter.CursorSequenceBuilder.build(), CachingClusteredClient.run(), etc. is a nightmare to read and I always feel frustration when I have to work with such methods.

I do agree with that (CachingClusteredClient.run is a real struggle) but that's 430 LOC with lots of futures, branches, and early returns. It's a really different story. In this case, the method is 46 LOC, fits on a screen, and is very simple. There's only one local class, and splitting it out would generate extra boilerplate (it would need a constructor).

If this method was longer, or if there was more branching and each branch had its own local class, I would probably agree with you and change it. But for this relatively simple method I can't see that breaking it up is objectively better than keeping it as-is. You're suggesting moving almost all of this code to a constructor of a new class, and replacing this method with a single line. It just seems different to me, not really better.

I think if you want a blanket rule that the local class pattern is bad and should be avoided no matter what (even for relatively simple methods) then that's a conversation we could have, but this PR isn't the right place to bring it up. It should be an addition to the code style guidelines.

Otherwise, if the project code style guidelines are silent on the matter, style should be up to the patch author as long as it's not very difficult to understand and as long as the general "implicit" style of the surrounding module is followed.